### PR TITLE
Account for array header alignment for multianewarray evaluator

### PR DESCRIPTION
**Pull request by hzongaro**
The array header size recently increased from 16 bytes to 24 bytes on 64-bit platforms.  With a 16 byte header size, inline code generated by the JIT to allocate a multidimensional array on the thread-local heap didn't need to account for alignment of the array headers, as they would be aligned for any shift compression amount.  With a size of 24 bytes,
headers will not be automatically aligned for a shift compression amount of four, which requires a heap alignment of 16 bytes.

The problem can be reproduced with this simple test:

```
public class Bar {
    public static final Integer[][] sub(int dim1, int dim2) {
        return new Integer[dim1][dim2];
    }

    public static final void main(String[] args) {
        sub(3, 0);
    }
}
```

Running with this command:

```
java -XXgc:forcedShiftingCompressionAmount=4 -Xjit:limit={Bar.sub*},count=0 Bar
```

Adjusted `multianewarray` evaluators for x and z platforms to account for array header alignment requirements.

### CODE REVIEW

**Comment by hzongaro** [comment]
Daryl @0xdaryl, Filip @fjeremic, may I ask you to review this change?

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -1307,9 +1311,9 @@ TR::Register *J9::X86::TreeEvaluator::multianewArrayEvaluator(TR::Node *node, TR
       generateRegImmInstruction(ANDRegImm4(), node, temp1Reg, -elementSizeAligned, cg);
       }
 
-   // temp2Reg = firstDimLenReg * 16 (discontiguousArrayHeaderSizeInBytes)
-   TR_ASSERT_FATAL(zeroArraySize >= 0 && zeroArraySize <= 127, "discontiguousArrayHeaderSizeInBytes cannot be > 127 for IMulRegRegImms instruction");
-   generateRegRegImmInstruction(IMULRegRegImm4(), node, temp2Reg, firstDimLenReg, zeroArraySize, cg);
+   // temp2Reg = firstDimLenReg * alignedDiscontiguousArrayHeaderSizeInBytes
```
Comment (0xdaryl) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp:1314]: Delete this commented line.

**Comment by 0xdaryl** [comment]
@dmitripivkine : FYI, this fix ensures the zero-length array header size is a multiple of the object alignment (16).  Please approve.

**Comment by fjeremic** [comment]
> The problem can be reproduced with this simple test:

@hzongaro can you please check in a unit test for this to ensure problems like this don't ever happen again in the future. For example I could see Power, AArch64, RISC-V, ARM, etc. coming in a year later and implementing acceleration for multinewarray and checking this exact same bug again. Adding a unit test will prevent that.

Please add it here with the already existing tests for multianewarray:
https://github.com/eclipse-openj9/openj9/blob/master/test/functional/JIT_Test/src/jit/test/jitt/array/multianewarray.java

You can use #11088 as an example of how to add a variation with the specific options you need to reproduce the issue.

**Comment by fjeremic** [comment]
@dchopra001 could you please review as the expert of the code in this area?

**Review by dmitripivkine — APPROVED** [review]
Looks like this is correct

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -3376,9 +3380,9 @@ J9::Z::TreeEvaluator::multianewArrayEvaluator(TR::Node * node, TR::CodeGenerator
...
 
    TR::Register *temp2Reg = cg->allocateRegister();
-   // temp2Reg = firstDimLenReg * 16 (discontiguousArrayHeaderSizeInBytes)
+   // temp2Reg = firstDimLenReg * alignedDiscontiguousArrayHeaderSizeInBytes
    generateRRInstruction(cg, TR::InstOpCode::LGR, node, temp2Reg, firstDimLenReg);
-   generateRILInstruction(cg, TR::InstOpCode::MSGFI, node, temp2Reg, zeroArraySize);
+   generateRILInstruction(cg, TR::InstOpCode::MSGFI, node, temp2Reg, zeroArraySizeAligned);
 
    cursor = generateRRInstruction(cg, TR::InstOpCode::AGR, node, temp2Reg, temp1Reg);
    iComment("Calculates (firstDimLen * zeroArraySize) + (arrayStrideInBytes + arrayHeaderSize)");
```
Comment (dchopra001) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:3388]: Can we replace the `zeroArraySize` here as well?

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -1263,7 +1263,11 @@ TR::Register *J9::X86::TreeEvaluator::multianewArrayEvaluator(TR::Node *node, TR
    // First Dim zero, only allocate 1 zero-length object array
    int32_t zeroArraySize = TR::Compiler->om.discontiguousArrayHeaderSizeInBytes();
    generateRegMemInstruction(LRegMem(), node, targetReg, generateX86MemoryReference(vmThreadReg, offsetof(J9VMThread, heapAlloc), cg), cg);
-   generateRegMemInstruction(LEARegMem(), node, temp1Reg, generateX86MemoryReference(targetReg, zeroArraySize, cg), cg);
+
+   // Take into account alignment requirements for the size of the zero-length array header
+   int32_t zeroArraySizeAligned = OMR::align(zeroArraySize, TR::Compiler->om.objectAlignmentInBytes());
```
Comment (dchopra001) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp:1268]: It seems like `zeroArraySize` is no longer used anywhere else except when we define `zeroArraySizeAligned` here. Can we just use `TR::Compiler->om.discontiguousArrayHeaderSizeInBytes()` and eliminate `zeroArraySize`? Similarly for Z.

**Review by dchopra001 — CHANGES_REQUESTED** [review]
Just a couple of minor nits. Otherwise looks good!

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp (hunk 2 above)
Comment (hzongaro) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:3388]: Yes - thanks for catching that!

Comment (hzongaro) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp]: Revised in commit e98874bf5b1be4188e76f129a472c4a0434cd525

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp (hunk 3 above)
Comment (hzongaro) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Revised in commit e98874bf5b1be4188e76f129a472c4a0434cd525

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp (hunk 1 above)
Comment (hzongaro) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Revised in commit e98874bf5b1be4188e76f129a472c4a0434cd525

**Comment by hzongaro** [comment]
Daryl @0xdaryl, Dhruv @dchopra001, I've addressed your comments in commit e98874bf5b1be4188e76f129a472c4a0434cd525.  If you're OK with things as they now stand, I'll squash those two commits in the compiler code into one.

**Review by fjeremic — APPROVED** [review]

**Comment by fjeremic** [comment]
Jenkins test sanity all jdk8,jdk11

**Review by dchopra001 — APPROVED** [review]

**Comment by 0xdaryl** [comment]
The Win32 CI failures look suspiciously relevant to this PR -> https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_x86-32_windows_Personal/28/tapResults/

**Comment by 0xdaryl** [comment]
Please squash the review commits as well.

**Comment by fjeremic** [comment]
> The Win32 CI failures look suspiciously relevant to this PR -> [openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_x86-32_windows_Personal/28/tapResults](https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_x86-32_windows_Personal/28/tapResults/)

```
JVMJ9GC040E Unknown option: forcedShiftingCompressionAmount=2
```

Looks like the GC is unable to handle this option on Windows 32-bit.

**Comment by hzongaro** [comment]
> Please squash the review commits as well.

Squashed.

> Looks like the GC is unable to handle this option on Windows 32-bit. You may have to exclude this platform in your variation.

I've split the new variations into their own separate test, requiring 64-bit and non-xl - I hope that will avoid all the cases for which `-XXgc:forcedShiftingCompressionAmount` is not supported.

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk8,jdk11

**Pull request merged**